Update GraalVM to 17.0.7 (23.0.0 JDK17)

Akirathan opened the pull request:
Closes #5300

Update GraalVM to **17.0.7+7.1**

### Pull Request Description
- Previous GraalVM update: https://github.com/enso-org/enso/pull/6750

Removed warnings:
- Remove deprecated `ConditionProfile.createCountingProfile()`.
- Add `@Shared` to some `@Cached` parameters (Truffle now emits warnings about potential `@Share` usage).
- Specialization method names should not start with execute
- Add limit attribute to some specialization methods
- Add `@NeverDefault` for some cached initializer expressions
- Add `@Idempotent` or `@NonIdempotent` where appropriate

BigInteger and potential Node inlining are tracked in follow-up issues.


### Important Notes

For `SDKMan` users:
```
sdk install java 17.0.7-graalce
sdk use java 17.0.7-graalce
```

For other users - download link can be found at https://github.com/graalvm/graalvm-ce-builds/releases/tag/jdk-17.0.7

Release notes: https://www.graalvm.org/release-notes/JDK_17/

R component was dropped from the release 23.0.0, only `python` is available to install via `gu install python`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] Do QA on Windows - make sure that the new GraalVM runtime is downloaded, installed, and found.
- [x] Ensure that `python` and `R` interop tests are successful.
- [x] Run and compare benchmarks.
- [x] The documentation has been updated, if necessary.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] Make sure benchmarks are OK

Commit:
Update GraalVM to 23.0.0 JDK17

Commit:
Update CHANGELOG

Commit:
Update build tool for new Graal release names

JaroslavTulach committed:
Merge branch 'develop' into wip/akirathan/5300-graalvm-update-23-00

JaroslavTulach commented on build.sbt:
```diff
@@ -20,8 +20,11 @@ import java.io.File
+val graalVersion          = "17.0.7"
+// Verwsion used for the Graal/Truffle related Maven packages
+val graalMavenVersion     = "23.0.0"
+val javaLangVersion       = "17"
```
```java
javaLangVersion = graalVersion.split("\\.")[0];
```
I think. Valid right now and potentially also valid in the future - if I understand the versioning scheme properly.

Akirathan replied:
Yes, it seems so.

JaroslavTulach commented on build/ci_utils/src/cache/goodie/graalvm.rs:
```diff
@@ -171,9 +159,30 @@ Java(TM) SE Runtime Environment Oracle GraalVM 17.0.7+8.1 (build 17.0.7+8-LTS-jv
+
+    #[test]
+    fn recognize_oneline_version() {
+        let version_line = "Java(TM) SE Runtime Environment Oracle GraalVM 17.0.7+8.1 (build 17.0.7+8-LTS-jvmci-23.0-b12)";
```
Got tricked, Pavel? If it says `Oracle GraalVM`, then it is not the open source version, I think. What is the URL that we are downloading?

Akirathan replied:
Yes, I got confused a bit. I think that I understand it now - The following output is from the "new EE version" (which is now free for testing purposes, this might be the modification, as I believe the older versions could only be downloaded via Oracle Technology Network?):
```
java version "17.0.7" 2023-04-18 LTS
Java(TM) SE Runtime Environment Oracle GraalVM 17.0.7+8.1 (build 17.0.7+8-LTS-jvmci-23.0-b12)
Java HotSpot(TM) 64-Bit Server VM Oracle GraalVM 17.0.7+8.1 (build 17.0.7+8-LTS-jvmci-23.0-b12, mixed mode, sharing)
```

And this is the output from the "new CE version":
```
openjdk version "17.0.7" 2023-04-18
OpenJDK Runtime Environment GraalVM CE 17.0.7+7.1 (build 17.0.7+7-jvmci-23.0-b12)
OpenJDK 64-Bit Server VM GraalVM CE 17.0.7+7.1 (build 17.0.7+7-jvmci-23.0-b12, mixed mode, sharing)
```

So nowadays, `Oracle GraalVM` == enterprise edition and `GraalVM CE` == community edition.

Akirathan replied:
I got "tricked" because the [official download page](https://www.graalvm.org/downloads/#) downloads the enterprise edition, and if you want to download the community edition, you have to click on "GraalVM Community Edition" under "Installation Options", which is not as intuitive as I would like.

Commit:
Do not use @CachedLibrary parameter in method behind TruffleBoundary

Commit:
Add .idea dir to gitignore in tools/enso4igv

Akirathan commented on tools/enso4igv/.gitignore:
```diff
@@ -1,2 +1,4 @@
 *.vsix
 nbcode
+!.idea
```
I believe this is needed so that `./run git-clean` does not remove the `.idea` directory here. I have been playing around with this Maven subproject, and it was more comfortable for me to open it in IDEA, rather than in VSCode.

JaroslavTulach replied:
OK.

Commit:
Improve .gitignore

So that run git-clean does not remove .vscode directory

Commit:
Fix gu program find

Commit:
GraalVM root directory name includes the build metadata identifier

Commit:
Fix native image build of engine-runner

Commit:
Remove deprecations

Commit:
Dispose context before System.exit

Commit:
Get rid of some warnings - Add @Shared to some @Cached parameters

JaroslavTulach commented on build/ci_utils/src/cache/goodie/graalvm.rs:
```diff
@@ -88,36 +79,37 @@ impl GraalVM {
-            // No Graal packages for Apple Silicon.
-            Arch::AArch64 if TARGET_OS == OS::MacOS => "amd64",
+            Arch::X86_64 => "x64",
             Arch::AArch64 => "aarch64",
```
Are you trying to fix #6862 by downloading `aarch64` on Mac M1? Fixing #6862 would be desirable, but I doubt one can fix that without having access to Mac M1 hardware.

Also, I'd like to avoid letting that issue block upgrade to GraalVM 23.0.0 JDK 17.

Akirathan replied:
No, sorry for the confusion. I rolled back this change in 83ca38475c51168ea7d1257d532e5a6208208d4e - the only change here is the renaming from `amd64` to `x64`. I won't touch M1-specific parts.

JaroslavTulach commented on build/ci_utils/src/cache/goodie/graalvm.rs:
```diff
@@ -166,14 +158,49 @@ mod tests {
-        let version_string = r"openjdk version 17.0.6 2023-01-17
-OpenJDK Runtime Environment GraalVM CE 22.3.1 (build 17.0.6+10-jvmci-22.3-b13)
-OpenJDK 64-Bit Server VM GraalVM CE 22.3.1 (build 17.0.6+10-jvmci-22.3-b13, mixed mode, sharing)";
+        let version_string = "openjdk version \"17.0.7\" 2023-04-18
```
How could we remove these hardcoded versions from here and reuse those already defined in `build.sbt`? Could we create a `sbt verify-proper-graalvm` task and invoke it on the downloaded JVM? Such an `sbt` task could easily re-use the version information from `build.sbt`.

Just an idea.

Akirathan replied:
I don't understand. This is just a Rust unit test that does not have anything to do with the actual version of GraalVM used, i.e, with the version specified in `build.sbt`.

JaroslavTulach commented on engine/runtime/src/bench/java/org/enso/interpreter/bench/benchmarks/semantic/NestedPatternCompilationBenchmarks.java:
```diff
@@ -38,7 +39,7 @@ public class NestedPatternCompilationBenchmarks {
         ctx = Context.newBuilder()
                 .allowExperimentalOptions(true)
-                .allowIO(true)
+                .allowIO(IOAccess.ALL)
```
I hope this is just a deprecation and the original method still remains in the API. Does it?

Akirathan replied:
Yes, this is just a deprecation, that was very straightforward to remove. The original method still remains in the API, but for how long?

Commit:
Suppress all the Node inlining warnings in runtime projects

Commit:
Suppress BigInteger message implement warning

Commit:
Get rid of warnings in runtime.

- Remove deprecated `ConditionProfile.createCountingProfile()`.
- Add `@Shared` to some `@Cached` parameters.
- Specialization method names should not start with execute
- Add limit attribute to some specialization methods
- Add `@NeverDefault` for some cached initializer expressions
- Add `@Idempotent` or `@NonIdempotent` where appropriate

Commit:
Merge branch 'develop' into wip/akirathan/5300-graalvm-update-23-00

# Conflicts:
#	engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/mutable/CoerceArrayNode.java

Commit:
Fix build.sbt after merge

Commit:
Add some `@Shared` parameters after develop merge

Commit:
Remove Truffle DSL processor options from projects that don't run the processor

Commit:
Rollback .idea run configuration changes

Commit:
Build tool: No Aarch64 graalvm package for Apple M1

jdunkerley commented on CHANGELOG.md:
```diff
@@ -851,6 +851,7 @@
 - [Added retries when executing GraalVM updater][7079]
 - [Add method call info for infix operators][7090]
 - [`executionComplete` response is sent on successful execution only][7143]
+- [Update to GraalVM 23.0.0][7176]
 - [Send info about function values][7168]
```
Why not in ascending order?

Akirathan replied:
Probably the default behavior of `git merge develop`. Reverted in 611212cf69922e0891863ac73f11da50ae1c0c27

jdunkerley commented on CHANGELOG.md:
```diff
@@ -975,6 +976,7 @@
 [7079]: https://github.com/enso-org/enso/pull/7079
 [7090]: https://github.com/enso-org/enso/pull/7090
 [7143]: https://github.com/enso-org/enso/pull/7143
+[7176]: https://github.com/enso-org/enso/pull/7176
 [7168]: https://github.com/enso-org/enso/pull/7168
```
As above.

Commit:
Fix order in CHANGELOG

Commit:
Remove every dir in PATH exists asserts from build script

Windows sets up unrelated PATHS that do not exist and therefore panics.

Akirathan commented on build/ci_utils/src/programs/graal.rs:
```diff
@@ -111,16 +110,17 @@ pub enum ComponentId {
     NodeJs,
     /// Graal.Python
     Python,
-    /// FastR
-    #[strum(serialize = "R")]
-    R,
```
I have removed FastR for now, as I am unable to install it or to list it with `gu available`. I asked at [Truffle tools slack](https://graalvm.slack.com/archives/C016UG9L5UP/p1688574032797399). Until that is resolved, we will have to skip all the R polyglot tests. Tracked in #7226.

Akirathan commented:
This PR is currently blocked on native image build failing on Windows with:
```
Error: On Windows, GraalVM Native Image for JDK 17 requires Visual Studio 2022 version 17.1.0 or later (C/C++ Optimizing Compiler Version 19.31 or later).
```
at https://github.com/enso-org/enso/actions/runs/5473230558/jobs/9966429269?pr=7176#step:10:4397

i.e. we first need to update the Visual Studio version on Windows workers.

<blockquote><img src="https://opengraph.githubassets.com/ac18e3e51d924ed219048c7350a0e32abea54e25b4669553af7364f5fd7fb7e1/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Update GraalVM to 17.0.7 (23.0.0 JDK17) · enso-org/enso@c5a892e</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

JaroslavTulach commented on engine/runtime-language-epb/src/main/java/org/enso/interpreter/epb/node/ContextRewrapNode.java:
```diff
@@ -92,7 +93,7 @@ long doWrappedLong(
       GuardedTruffleContext origin,
       GuardedTruffleContext target,
-      @CachedLibrary(limit = "5") InteropLibrary numbers) {
+      @Shared("interop") @CachedLibrary(limit = "5") InteropLibrary numbers) {
```
Is it good idea to share `numbers` with `bools`?

Akirathan replied:
I am not sure about that. But in c98693780626cca8ee3ab4b41e0bc52a6c6598a4 I have reverted so that `bools` have their own `@Exclusive` InteropLibrary.

JaroslavTulach commented on engine/runtime-language-epb/src/main/java/org/enso/interpreter/epb/runtime/PolyglotProxy.java:
```diff
@@ -80,9 +83,9 @@ void leaveOrigin(InteropLibrary node, Object prev) {
-      @CachedLibrary(limit = "5") InteropLibrary errors,
-      @Cached @Cached.Exclusive ContextRewrapExceptionNode contextRewrapExceptionNode,
-      @Cached @Cached.Exclusive BranchProfile profile) {
+      @Shared("interop") @CachedLibrary(limit = "5") InteropLibrary errors,
```
Please note there is `InteropLibrary` for `nulls` and another one for `errors` - e.g. they are intentionally separated. Are benchmarks OK even after all this `@Shared` annotation adding?

Akirathan replied:
There is only `@CachedLibrary("this.delegate") InteropLibrary null`. The only dispatched library here is the one for `errors` that is shared, and used only in the exception branch.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -572,10 +584,10 @@ Object doConvertDate(
-      @Cached MethodResolverNode methodResolverNode) {
+      @Shared("typesLib") @CachedLibrary(limit = "10") TypesLibrary types,
+      @Shared("interop") @CachedLibrary(limit = "10") InteropLibrary interop,
+      @Shared("warnsLib") @CachedLibrary(limit = "10") WarningsLibrary warnings,
```
Wouldn't it be better to keep the name of argument and string send to `@Shared` in sync? E.g. use `@Shared("warnings")`?

Akirathan replied:
Renamed in 7118056bb79aeb0a7e25790f7304e5ce25b34ab2.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -621,6 +634,7 @@ Object doConvertDateTime(
     }
   }
 
+  @NonIdempotent
```
What an [@NonIdempotent annotation](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/dsl/NonIdempotent.html)!

What is preventing this specialization to be idempotent?

Akirathan replied:
This annotation should not have been here. It seems like a leftover of my previous endeavors.

Akirathan replied:
Removed in 1cd76f9bf4c1044f17e77e704a5380c8ec9d11cd

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/io/PrintlnNode.java:
```diff
@@ -91,17 +93,20 @@ private void print(PrintStream out, Object str) {
         InvokeCallableNode.ArgumentsExecutionMode.PRE_EXECUTED);
   }
 
+  @NeverDefault
```
Another [interesting @NeverDefault annotation](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/dsl/NeverDefault.html).


I am not sure why it is needed for a method that constructs a `Node` when constructing a `Node` is forbidden on fast path?

Akirathan replied:
`PrintlnNodeGen.create()` is `@NeverDefault`, but it seems that Truffle cannot determine that at this point, so one has to annotate even this `build()` method. We might avoid these complications by following Truffle's naming conventions for `create` methods.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/function/Function.java:
```diff
@@ -141,6 +142,7 @@ public MaterializedFrame getScope() {
    *
    * @return the function's argument schema
    */
+  @Idempotent
```
Finally an idempotent function!

JaroslavTulach commented on test/Tests/src/Network/Http_Spec.enso:
```diff
@@ -309,7 +309,7 @@ spec =
                     "Content-Length": "16",
                     "Content-Type": "application/json",
-                    "User-Agent": "Java-http-client/17.0.6"
+                    "User-Agent": "Java-http-client/17.0.7"
```
I suggest to replace line:
```ruby
res.body.decode_as_json.should_equal expected_response
```
with
```ruby
simpleBody = res.body.replace_all 'User-Agent: *Java-http-client/.*$' 'User-Agent: Java-http-client`
simpleBody.decode_as_json.should_equal expected_response
```

Akirathan replied:
Good idea. But I am struggling too much now with this replacement. It does not work out of the box and I don't want to spend too much time on that. Leaving that one be for now.

JaroslavTulach reviewed (approved):
Looks promissing:
- make sure our benchmarks run fast
- work with @mwu-tow to pass the Windows build

Commit:
Bools and numbers do not share interop lib

Commit:
Remove `@NonIdempotent` annotation from specialization

Commit:
Rename shared cached parameter names

Commit:
Remove project/EnvironmentCheck

It is no longer needed - Java and GraalVM have the same version nowadays.

Commit:
Merge branch 'develop' into wip/akirathan/5300-graalvm-update-23-00

# Conflicts:
#	CHANGELOG.md
#	engine/runtime/src/main/java/org/enso/interpreter/node/callable/argument/ReadArgumentNode.java

Commit:
Remove @CachedLibrary parameters from methods behind @TruffleBoundary

Commit:
Fix GraalVM version handling in project-manager

Commit:
fmt

Commit:
java version is specified as SemVer

Commit:
Add version requirements to GraalVMVersion

xvcgreg commented:
@Akirathan what's the holdup here? Is it ready for merge?

Akirathan commented:
> @Akirathan what's the holdup here? Is it ready for merge?

I am still dealing with problems with different versioning of GraalVM introduced since the newest version 23.0.0. We are dependent on the versioning scheme at least in our build scripts, `build.sbt` projects, and project-manager. After resolving these problems, I will have to run benchmarks manually and ensure there are no regressions. After that, we are ready for merge.

Commit:
Fix enso-build-cli GraalVM version handling

Commit:
Loose requirements for GraalVMVersion

Commit:
Do not use invalid GraalVM version in tests

Commit:
fmt

JaroslavTulach commented:
@mwu-tow we are having issues with Rust based build system and need your review and guidance.

- [build.sbt needs formatting](https://github.com/enso-org/enso/actions/runs/5590783151/jobs/10220966862?pr=7176#step:5:140) - let's run `sbt scalafmtSbt`
- [lints in Rust](https://github.com/enso-org/enso/actions/runs/5590783140/jobs/10220968480?pr=7176) - let's run `cargo fmt` frequently
- [Illegal character " on Windows only](https://github.com/enso-org/enso/actions/runs/5590783155/jobs/10220967627?pr=7176#step:10:1093)!? - what can that mean when the [linux build is OK](https://github.com/enso-org/enso/actions/runs/5590783155/jobs/10220967419?pr=7176)?
- [mac os x failure](https://github.com/enso-org/enso/actions/runs/5590783155/jobs/10220967182?pr=7176) is probably normal and unrelated to this PR

I am going to fix what I can. Let's close this issue, finally!

JaroslavTulach committed:
Formatting build.sbt

JaroslavTulach committed:
Formatting scala sources

JaroslavTulach committed:
cargo fmt

JaroslavTulach committed:
Merge branch 'wip/akirathan/5300-graalvm-update-23-00' of enso:enso-org/enso into wip/akirathan/5300-graalvm-update-23-00

JaroslavTulach commented:
> After resolving these problems, ... run benchmarks manually and ensure there are no regressions. 

The Linux build is OK, so I have just started the [benchmarks run](https://github.com/enso-org/enso/actions/runs/5595448670).

Akirathan commented:
> The Linux build is OK, so I have just started the [benchmarks run](https://github.com/enso-org/enso/actions/runs/5595448670).

The [benchmark run](https://github.com/enso-org/enso/actions/runs/5595448670) failed after 4 minutes during compilation. It seems like the bench node was not properly cleaned. I tried triggering another run - https://github.com/enso-org/enso/actions/runs/5596132113

<blockquote><img src="https://opengraph.githubassets.com/2450f0e0b0d2cd139a98f2b68b17cd44f024708c5b440078e16ebe8e7370dbb1/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Benchmark Engine · enso-org/enso@013e161</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

Commit:
Remove @Shared warning in ReadArgumentCheckNode

Akirathan commented:
The failure in the [Windows Engine test](https://github.com/enso-org/enso/actions/runs/5590783155/jobs/10220967627?pr=7176#step:10:1093) seems to be caused by `sbt` wrong handling of Truffle annotation processor output. Although not a real fix, but https://github.com/enso-org/enso/pull/7176/commits/c66e3bd0be8825820c475455ecdef092a30d65ec removes the warning in `ReadArgumentNode` that caused this failure.

JaroslavTulach commented:
> caused by `sbt` wrong handling of Truffle annotation processor output. 

`sbt` and annotation processors - never ending problems. Does it make sense to report an error to `sbt` guys (or guy), @hubertp? We can just point to revision in history of our Git repository and the issue is going to reproduce itself immediately, right?

JaroslavTulach commented:
Windows build fails on [cannot find package graalvm-ce-java11-windows-amd64-1.0.0.zip](https://github.com/enso-org/enso/actions/runs/5597347620/jobs/10235574414?pr=7176#step:10:7444).

Akirathan commented:
> Windows build fails on [cannot find package graalvm-ce-java11-windows-amd64-1.0.0.zip](https://github.com/enso-org/enso/actions/runs/5597347620/jobs/10235574414?pr=7176#step:10:7444).

At first sight, this seems like another instance of dirty-caches-issue. In the last few commits, I renamed a lot of files inside `runtime-version-manager-test/src/main/resources` to satisfy the new versioning requirements. This changed might not been propagated to the windows runners. There was a recent clean action on the CI, so I am restarting the job now and hopefully, it will work.

Akirathan commented:
I have just checked the benchmarks, and there are no big regressions, neither improvements anywhere. Seems like the performance stays roughly the same. Which is precisely the result that we wished for.

radeusgd commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/components/RuntimeVersionManager.scala:
```diff
@@ -654,7 +655,7 @@ class RuntimeVersionManager(
   private def parseGraalRuntimeVersionString(
     name: String
   ): Option[GraalVMVersion] = {
-    val regex = """graalvm-ce-java(\d+)-(.+)""".r
+    val regex = """graalvm-ce-java(.+)-(.+)""".r
```
This change is itself backwards compatible, but that means newer engine versions will be accepting Java versions (like this one we are upgrading into) that were not supported by the old `runtime-version-manager`.

This means that the launcher/project-manager will fail when it sees an engine with such a problematic runtime:

```
[error] [2023-07-19T14:13:35.009Z] [org.enso.launcher.cli.Main] A fatal error has occurred: The runtime GraalVM 23.0.0 Java 17.0.7 is already installed, but cannot be loaded due to Invalid runtime component name `graalvm-ce-java17.0.7-23.0.0`.. Until the launcher gets an auto-repair feature, please try reinstalling the runtime by uninstalling all engines that use it and installing them again, or manually removing `C:\Users\progr\AppData\Local\enso\runtime\graalvm-ce-java17.0.7-23.0.0`. (Caused by: Invalid runtime component name `graalvm-ce-java17.0.7-23.0.0`.)
```

While I know the `launcher` project is not used too much (sadly, it's really useful), I'm not sure how big of a priority it is to make sure this work. However, this also affects the `project-manager` - it will have a similar error when old IDE encounters such a runtime.

If we change the `manifest.template.yaml`:
```diff
-minimum-launcher-version: 0.2.13
-minimum-project-manager-version: 0.2.13
+minimum-launcher-version: 2023.2.1-nightly.2023.7.20
+minimum-project-manager-version: 2023.2.1-nightly.2023.7.20
```
It will fix that issue.

The launcher will then propose to upgrade to the tomorrows nightly (won't work today but that's a short time period):
```
[warn] [2023-07-19T14:40:38.253Z] [org.enso.launcher.upgrade.LauncherUpgrader.auto-upgrade] A more recent launcher version (at least 2023.2.1-nightly.2023.7.20) is required to continue.
Do you want to upgrade the launcher and continue? [Y/n] 
```

Similarly, the `project-manager` should display a slightly more friendly error message in the logs saying that to open this project a more recent version of `project-manager` is needed.

I think we should update the manifest to reflect these breaking changes and ensure that old launcher/project-manager don't crash and burn but display nice error messages.

Ideally, we should set the version to be some full release, not a nightly - we can bump this with a PR once we have such a release.

Akirathan replied:
I bumped the version in 881c98378864df8e7f3e6f06550819c07e6dee5f. Let's see if something fails now.

radeusgd committed:
Add a missing placeholder for a runtime-version-manager-test

radeusgd commented on lib/scala/runtime-version-manager/src/main/scala/org/enso/runtimeversionmanager/releases/graalvm/GraalCEReleaseProvider.scala:
```diff
@@ -20,17 +20,22 @@ class GraalCEReleaseProvider(releaseProvider: SimpleReleaseProvider)
-    val tagName     = s"vm-${version.graalVersion}"
     val packageName = packageFileName(version)
-    val release     = releaseProvider.releaseForTag(tagName)
+    val tagName = if (version.graalMajorVersion < 23) {
+      s"vm-${version.graalVersion}"
+    } else {
+      s"jdk-${version.javaVersion}"
+    }
+    val release = releaseProvider.releaseForTag(tagName)
```
We now have a new codepath that depends on Graal major version being less or greater than `23`.

But our unit tests only check Graal versions `1.0.0` and `2.0.0` both with "Java 11". Please add a test that has Graal major version >= 23 and uses the new Java format, e.g. `17.0.7` or something.

Akirathan replied:
Added in 93f504c085114ca26782f38fb670303cf6d137f7

Commit:
Add RuntimeVersionManager test for the new GraalVM versioning

Commit:
Bump minimal versions in manifest template

Commit:
fmt

Commit:
Fix EngineManagementApi project-manager test

Commit:
Merge branch 'develop' into wip/akirathan/5300-graalvm-update-23-00

# Conflicts:
#	build.sbt

mwu-tow commented on build/ci_utils/src/cache/goodie/graalvm.rs:
```diff
@@ -166,14 +179,55 @@ mod tests {
-        let version_string = r"openjdk version 17.0.6 2023-01-17
-OpenJDK Runtime Environment GraalVM CE 22.3.1 (build 17.0.6+10-jvmci-22.3-b13)
-OpenJDK 64-Bit Server VM GraalVM CE 22.3.1 (build 17.0.6+10-jvmci-22.3-b13, mixed mode, sharing)";
+        let version_string = "openjdk version \"17.0.7\" 2023-04-18
```
Why remove raw-string literal? Now I cannot just copy-paste output, as the quotes need to be escaped.

Akirathan replied:
I was not aware how to do raw string literal with double quotes inside. But now I have checked that out and fixed that in 7389e4191986dbc49916ab9a64af46bee300c1a0

mwu-tow commented on build/ci_utils/src/cache/goodie/graalvm.rs:
```diff
@@ -54,27 +52,40 @@ impl Goodie for GraalVM {
 
     fn activation_env_changes(&self, package_path: &Path) -> Result<Vec<crate::env::Modification>> {
-        let package_path = package_path.join(self.root_directory_name());
+        let dir_entries = package_path
+            .read_dir()
+            .context("Failed to read GraalVM cache directory")
+            .unwrap()
+            .collect_vec();
+        assert_eq!(
+            dir_entries.len(),
+            1,
+            "GraalVM cache directory should contain exactly one directory"
+        );
+        let graalvm_dir = match dir_entries.get(0).unwrap() {
+            Ok(dir_entry) => dir_entry,
+            Err(err) => bail!("Failed to read GraalVM cache directory: {}", err),
+        };
+        let dir_name_tmp = graalvm_dir.file_name();
+        let dir_name = dir_name_tmp.as_str();
+        assert!(dir_name.contains("graalvm"));
+        let graal_version_str = format!(
+            "{}.{}.{}",
+            self.graal_version.major, self.graal_version.minor, self.graal_version.patch
+        );
+        assert!(dir_name.contains(&graal_version_str));
```
In general, we never want to use `unwrap` or `assert` in non-test, as it causes a hard panic.

Instead of `unwrap` use `?`, so the function returns error.
Similarly, in place of `assert!` use `ensure!` with an appropriate message.
To make sure that there is only one element in vector (and retrieve it) you can do something along lines:
```
    let [graalvm_dir] = dir_entries.as_slice() else { bail!("GraalVM cache directory should contain exactly one directory") };
```
(also, it'd be nice to include the relevant path in the error message)

Akirathan replied:
Got it. Tried to fix it in 2a1f553404518fdfe312f3c29dd5900c7d8f43c1

mwu-tow commented on build/ci_utils/src/cache/goodie/graalvm.rs:
```diff
@@ -54,27 +52,40 @@ impl Goodie for GraalVM {
+        let dir_name_tmp = graalvm_dir.file_name();
+        let dir_name = dir_name_tmp.as_str();
+        assert!(dir_name.contains("graalvm"));
+        let graal_version_str = format!(
```
Why not just `self.graal_version.to_string()` (unless we want to explicitly reject patch/prerelease pieces — but I don't see them in build.sbt?)
If we truly wanted to do this, I'd rather see this as a method in `VersionExt` (like `to_string_core` por sth), so the intent is clearly expressed.

Akirathan replied:
`graal_version_str` is only used for `assert!(dir_name.contains(&graal_version_str))`. The thing is that when you download the GraalVM distribution nowadays, and extract it, the directory name in the extracted folder has a weird name that includes the build metadata. And I have not found an easy way how to infer the directory name here. So I am just assuming that there is just one directory (as there has always been), and do some sanity checks on the name of the directory so that I am sure that the downloaded tarball contains the directory with graalvm.

Nevertheless, I added `to_string_core` method and its usages in 57c37ed3f08d7c0b07a14ba9aa6ccaeb8b07c0c5

mwu-tow commented on build/ci_utils/src/cache/goodie/graalvm.rs:
```diff
@@ -54,27 +52,40 @@ impl Goodie for GraalVM {
+            Ok(dir_entry) => dir_entry,
+            Err(err) => bail!("Failed to read GraalVM cache directory: {}", err),
+        };
+        let dir_name_tmp = graalvm_dir.file_name();
```
It's not really a temp, if (as I suppose) you added this to make a result of temporary expression non-temporary.
I'd just name this `dir_name` and shadow the name next. The value will live on, unaffected.

Akirathan replied:
Yes, that was exactly the purpose. Fixed in 2064a5ca162fd6ba02b4e4c7f1d354b403732f8b by shadowing the `dir_name` variable.

mwu-tow commented on build/ci_utils/src/cache/goodie/graalvm.rs:
```diff
@@ -88,36 +99,37 @@ impl GraalVM {
     pub fn url(&self) -> BoxFuture<'static, Result<Url>> {
         let platform_string = self.platform_string();
         let graal_version = self.graal_version.clone();
+        let graal_version_tag =
+            format!("{}.{}.{}", graal_version.major, graal_version.minor, graal_version.patch);
```
As above.

Akirathan replied:
Fixed in 57c37ed3f08d7c0b07a14ba9aa6ccaeb8b07c0c5

mwu-tow commented on build/ci_utils/src/cache/goodie/graalvm.rs:
```diff
@@ -88,36 +99,37 @@ impl GraalVM {
-
-    pub fn root_directory_name(&self) -> PathBuf {
-        PathBuf::from(format!("{}-{}-{}", PACKAGE_PREFIX, self.java_version, self.graal_version))
+        let java_version = format!(
+            "jdk-{}.{}.{}",
+            _graal_version.major, _graal_version.minor, _graal_version.patch
+        );
```
as above

Akirathan replied:
Fixed in 57c37ed3f08d7c0b07a14ba9aa6ccaeb8b07c0c5

mwu-tow commented on build/ci_utils/src/cache/goodie/graalvm.rs:
```diff
@@ -166,14 +179,55 @@ mod tests {
+            pre:   Prerelease::EMPTY,
+            build: BuildMetadata::new("7.1").unwrap(),
+        };
+        let version_str = format!(
```
As above. But then — I'm not sure if we need such test, as it'd end up testing `semver` library.

Akirathan replied:
Right, this test is just a leftover when I was fiddling with the versions. Removed in 0a0545ef71140d754ec539a5d6d1cc6cc9345bca

Commit:
Use raw string literal

Commit:
Do not use new variable name just to hold temporary expression.

Commit:
Implement VersionExt.to_string_core

Commit:
Remove unnecessary test.

It tested just semver library.

Commit:
Do not use assert! in non-tests

mwu-tow commented on build/ci_utils/src/cache/goodie/graalvm.rs:
```diff
@@ -54,27 +52,33 @@ impl Goodie for GraalVM {
+        let [graalvm_dir] = dir_entries.as_slice() else {
+            bail!("GraalVM cache directory should contain exactly one directory");
+        };
+        let graalvm_dir = match graalvm_dir {
```
Something like
```
let graalvm_dir = graalvm_dir.context("Failed to read GraalVM cache directory.")?;
```

Would be a little bit more idiomatic (context adds additional error "layer", we don't need then to manually merge it with previous error.

Akirathan replied:
For now, I decide to ignore this suggestion, as after applying, the borrow checker fails with "cannot move out of `*graalvm_dir` which is behind a shared reference". And I don't really want to bother with this right now. Thanks for the explanation anyway.

Commit:
fix linter errors

Commit:
Remove @NonIdempotent annotation from specializations
